node/src/components/deploy_acceptor/tests.rs

```diff
@@ -0,0 +1,584 @@
+                    source,
+                    responder: Some(responder),
+                },
+                QueueKind::Regular,
+            )
+            .ignore()
+    }
+}
+
+async fn run_deploy_acceptor_without_timeout(
+    test_scenario: TestScenario,
+) -> Result<(), super::Error> {
+    let mut rng = crate::new_rng();
+
+    let mut runner: Runner<ConditionCheckReactor<Reactor>> =
+        Runner::new(test_scenario, &mut rng).await.unwrap();
+
+    let block = Box::new(Block::random(&mut rng));
+    // Create a responder to assert that the block was successfully injected into storage.
+    let (block_sender, block_receiver) = oneshot::channel();
+    let block_responder = Responder::create(block_sender);
+
+    runner
+        .process_injected_effects(put_block_to_storage(block, block_responder))
+        .await;
+
+    // There's only one scheduled event, so we only need to try cranking until the first time it
+    // returns `Some`.
+    while runner.try_crank(&mut rng).await.is_none() {
+        time::sleep(POLL_INTERVAL).await;
+    }
+    assert!(block_receiver.await.unwrap());
+
+    // Create a responder to assert the validity of the deploy
+    let (deploy_sender, deploy_receiver) = oneshot::channel();
+    let deploy_responder = Responder::create(deploy_sender);
+
+    // Create a deploy specific to the test scenario
+    let deploy = test_scenario.deploy(&mut rng);
+    // Mark the source as either a peer or a client depending on the scenario.
+    let source = test_scenario.source(&mut rng);
+
+    {
+        // Inject the deploy artificially into storage to simulate a previously seen deploy.
+        if test_scenario == TestScenario::FromClientRepeatedValidDeploy
+            || test_scenario == TestScenario::FromPeerRepeatedValidDeploy
+        {
+            let injected_deploy = Box::new(deploy.clone());
+            let (injected_sender, injected_receiver) = oneshot::channel();
+            let injected_responder = Responder::create(injected_sender);
+            runner
+                .process_injected_effects(put_deploy_to_storage(
+                    injected_deploy,
+                    injected_responder,
+                ))
+                .await;
+            while runner.try_crank(&mut rng).await.is_none() {
+                time::sleep(POLL_INTERVAL).await;
+            }
+            // Check that the "previously seen" deploy is present in storage.
+            assert!(injected_receiver.await.unwrap());
+        }
+    }
+
+    runner
+        .process_injected_effects(schedule_accept_deploy(
+            Box::new(deploy.clone()),
+            source,
+            deploy_responder,
+        ))
+        .await;
+
+    // Tests where the deploy is already in storage will not trigger any deploy acceptor
+    // announcement, so use the deploy acceptor `PutToStorage` event as the condition.
+    let stopping_condition = move |event: &Event| -> bool {
+        match test_scenario {
```

Comment:
Similar to the above, I'd rather see this logic hidden inside the `TestScenario` implementation. Consider adding a function to the `TestScenario` impl, for example `fn is_stop_condition(&self, event: &Event) -> bool` and move the match there.
On the caller site you can just use:

```
    runner
        .reactor_mut()
        .set_condition_checker(Box::new(move |event: &Event| -> bool {
            test_scenario.is_stop_condition(event)
        }));
```

Reply:
So both in this case and the above case, I want to have this present in the deploy acceptor function itself, cause I feel its better to define the stopping conditions for the `deploy_acceptor` in this function itself, as opposed to have it extracted away into a utility enum. Moreover, I think this makes it more consistent as it follows the same pattern in the gossiper tests. I am not going to plant my flag on this issue, however. So if @Fraser999 over-rules me here, I am okay with extracting them away